### PR TITLE
[SPARK-25857][core] Add developer documentation regarding delegation tokens.

### PR DESCRIPTION
**Pull request by vanzin**

### CODE REVIEW

**Comment by vanzin** [comment]
@tgravescs @jerryshao

**Review by HyukjinKwon — COMMENTED** [review]
Looks fine to me otherwise. I did a quick proof reading.

**Comment by HyukjinKwon** [comment]
It doesn't block this PR but how about considering to put this in somewhere like https://spark.apache.org/docs/latest/running-on-yarn.html after refining it? We have Kerberos chapter https://spark.apache.org/docs/latest/running-on-yarn.html#kerberos

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: `For this reason, certain DT
providers cannot provide a renewal period to the Spark code`
Just for the understanding are these HBase + Hive, right?

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: Maybe worth to mention the provider enable flag?

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: `Without access to Kerberos credentials,
Spark cannot create DTs`
One exception is Kafka, which can get token with SSL certificate authentication.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Linux OS users and not kerberos users.

**Review by gaborgsomogyi — COMMENTED** [review]
High quality doc, thanks for creating it.
Found some minor stuff but looks good.

**Comment by gaborgsomogyi** [comment]
@HyukjinKwon do you mean here copy or link there?
`It doesn't block this PR but how about considering to put this in somewhere like https://spark.apache.org/docs/latest/running-on-yarn.html after refining it? We have Kerberos chapter https://spark.apache.org/docs/latest/running-on-yarn.html#kerberos`

**Comment by vanzin** [comment]
> how about considering to put this in somewhere like 

This is not for consumption by users. It's for consumption by Spark developers who want to mess with the delegation token functionality inside Spark.

I have to explain this stuff every time someone tries to use delegation tokens somewhere in Spark, so now I can just point to this doc.

**Comment by gaborgsomogyi** [comment]
> This is not for consumption by users. It's for consumption by Spark developers who want to mess with the delegation token functionality inside Spark.

+1 and this doc is not yarn specific.
I was thinking about a link maybe but started to think it's not good idea.

**Review by srowen — COMMENTED** [review]
I think this could reasonably go in the main Spark docs too, but up to you. I know it's long and fairly niche.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: This document is in the same package as the API. It's targeted at developers changing / using that API.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: I think listing all services that use DTs is kinda pointless. Otherwise, let's also includ Solr, Kudu, and who knows what else.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: There's an example in the next topic if this is not clear. This is the naming used by `kinit` (see its man page).

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: That's the case in the current providers available in Spark. But I can't say those are the only examples. This is not meant to explain the Spark code, it's meant to explain DTs in general.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: This is not user documentation, so not sure what benefit that would bring.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: I see. How about adding the class name `HadoopDelegationTokenProvider` or explicitly mentioning it's an internal API? Maybe i'm too much caring but wonder if it could give an impression that we have an external API that support DT creation.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: not a bit deal but I think we better stick to one way. " Key Distribution Center (KDC)" vs "TGT (ticket granting ticket)".

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (jerryshao) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Yes, I think we should differentiate renew and recreate, AFAIK in the Spark side we always create a new DT before the renewal interval.

**Comment by HyukjinKwon** [comment]
re: https://github.com/apache/spark/pull/23348#issuecomment-451229540

Yea, the current status looks closer to dev docs. But likewise I had to research and explain this stuff to users as well :-). Should be good to document in main doc somewhere after refining but yea let me leave it to you. Looks find to me anyway.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: The point here not the list but the fact there are services [which are] + [which aren't] in the Hadoop ecosystem. It's a minor thing so no strong opinion.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Here I didn't expect any changes, just wanted to know which are fulfilling the criteria.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: The names are fine. AFAIK after lifetime reached the token can't be renewed but has to be recreated.

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: > after lifetime reached the token can't be renewed 

No, that's the renewable life.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: I'll just revert to KDC. Readers of this document are expected to at least be a little familiar with Kerberos, so those terms should be familiar.

**Comment by vanzin** [comment]
> explain this stuff to users as well 

It's fine to point users at this if you want. But it's not meant to be user documentation, because it doesn't provide anything actionable for users.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Yeah, written the wrong name. The point is such case worth to say recreate because renew is misleading.

**Review by gaborgsomogyi — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: this isn't true, delegation tokens don't allow for single place.  you could just as well have a keytab in the single place that distributes the tgts. And really its not a single place since it distributes them to executors.  We might rephrase this as an implementation detail that they go to the driver and are distributed

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: remove extra newline

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: I think this is a yarn only way of distributing, @vanzin k8s, mesos doesn't use UGI does it?

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: perhaps you should use max lifetime and expiration time, it has to be renewed before the expiration time but it can only be renewed up to the max lifetime

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: I believe hive ones are also renewed.  I don't remember if hbase ones need renewed or are just a single lifetime. I think my point is this is a pretty broad statement that could be incorrect so I think we should change the wording.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: while this is true I'm not sure this is an issue, perhaps we should just point out that different services implement them differently, not all of them need a renewal.    ie your first sentence "to make matters worse" I think should be removed.  In a way its nice that hbase tokens don't need renewal as you don't have to worry about that in your code.  But it does lead to inconsistency between services.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: I think we should explicitly state the security ramifications of this.  distributed a keytab means that if its compromised someone can get tokens for any service as you.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: We seem to be mixing the section of what is a delegation token with what the benefits are, the definition of what also seems to be in  the next section with the lifecycle, perhaps we need to reorganize this a bit to have more of a section on what is DT, within there you can go into the benefits or why

Might mention here that with tokens you can't acquire more tokens vs with tgts or keytab you can get more tokens so its more secure to use tokens.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: > recreate because renew is misleading

It's not. Here you *can* renew the token. The "renewable life" controls what you're saying.

> perhaps you should use max lifetime and expiration time

Again, this is the same naming as kerberos documentation will give you, which I think is, in the end, less confusing. I'll use HDFS names also, just in case.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: >  delegation tokens don't allow for single place

Hmm, yes they do. Yes, you could do those other things, but then you wouldn't need DTs. So DTs are one way to allow a single place to handle kerberos credentials.

>  delegation tokens don't allow for single place

Not sure I follow. Only a single place has *kerberos* credentials. DTs are distributed to executors, but they are not *kerberos* credentials.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: The intent here is explained in the following paragraph. I'll add "per service" but I don't see how any other rephrasing can replace the following paragraph.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: It explains why Spark needs so much code to deal with delegation tokens, including an extensible interface, that wouldn't be needed if they were more like kerberos (where you just talk to one place).

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: You're thinking about writing the tokens to a file and automatically loading them, which is a YARN-only thing. Spark has its own mechanism to distribute tokens, which it uses even in YARN.

The YARN token API is mostly used so YARN can use the app's tokens for log aggregation and stuff.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: YARN does not have Hive client libraries in its classpath, so how can it even talk to the HMS or HS2 at all?

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: This is build-up for explaining why Spark handles DTs the way it does.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Explain more than "it's like writing your password to a file"?

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: ok checked with our hive team I was wrong.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: I'm not following the sentence above says spark distributes tokens to the executors using the UGI hadoop api. this works on yarn because its passed when you launch an executor, but how does the UGI get from the driver to executors on mesos?

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Oh I think I see what you are trying to say here.  I think you need to elaborate more on this.  I think you are trying to say that if you are sending kerberos credentials like a keytab that allows us to do all DT handling in a single place - the driver.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: ok, I guess I don't see the benefit in talking about what we could do, I would prefer to see it just state the drawback and if you want to compare to kerberos, compare to it.  But its not that big of a deal either.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Yes, I can see people not understanding that is a security concern.   If the file is written to HDFS with permissions restrictive, that might seem ok, but many security people would disagree.  Perhaps just add another sentence talking about this is a security concern because if the keytab is compromised people can get tokens for any service for an indefinite amount of time.  So users should evaluate if using this approach is ok.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: No, what I'm trying to say is that when you use DTs, there's a single place that needs to handle kerberos credentials.

It doesn't matter if you're using keytabs or kinit. It's still a single place. And that single place is responsible for distributing the DTs to the other entities that needs to authenticate against services.

It also doesn't need to be the Spark driver. As I mention in other places in the document, the DTs can be created by an external entity (like Oozie does).

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: ok, I'll remove.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Spark distributes the tokens using its RPC system, but they're stashed in the UGI, so if an executor needs to get the tokens, it needs to look at the UGI.

Anyway, I'll replace "distributes" since it seems to be causing confusion.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: It sounds like you want to warn users of how dangerous this can be. This is not end-user documentation, so even if I make that statement here, it will probably not have the effect you want.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: ok, sorry I understand now, I somehow really misread this.

**Comment by tgravescs** [comment]
one minor comment, otherwise +1, sorry again about my confusion yesterday on the one section.

**Comment by tgravescs** [comment]
+1

**Comment by vanzin** [comment]
Merging to master.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md:214]: @tgravescs Since not many peoples are dealing with security maybe just you and me so asking you directly. I've just double checked the code because wanted to use the mentioned feature but I have the feeling it's not working. After some digging I've not seen any significant code changes in the security area so I have the feeling it's maybe never worked. Have you ever looked after this feature? When I take a look at the current code I see that the obtained tokens are overwriting the existing ones.

Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md:214]: As far as I can remember this worked at one point and time, but there has been a lot of changes to the security code since I possibly tried it. I don't really deal with security anymore so don't have a setup to try.

Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md:214]: Good to hear at least once it was working. Do you know by heart roughly where to look for the mentioned filtering?
I though I would see some if condition not to set UGI but there is no such code.

Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md:214]: Sounds like the problem is that it picks them up by then overwrites?  I would look for the calls to addCredentials, I don't remember the details on that api to know if they overwrite or merge things. Seems like at one point we specifically handled merging but like I said its been a long time since I wrote that code and many people changed it since.

Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md:214]: The HADOOP_TOKEN_FILE_LOCATION stuff was just all handled by Hadoop for us and we inherit so might not see any code directly related to that. If its overwriting them, my guess is we are just smashing in the tokens rather then dealing with them there.

Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md:214]: Thx for the pointers, going to sit together with the hadoop guys and having a look at HADOOP_TOKEN_FILE_LOCATION + addCredentials in depth. AFIAR addCredentials overwrites by default but will see... Thx again!

**Pull request closed without merging**
